uefi: use tpm2-tss library to read PCR values

diabonas opened the pull request:
As discussed in #1307, directly use the tpm2-tss library to read PCR values instead of calling an external tpm2-tools binary. There are some issues with this WIP implentation:

- [ ] Currently only PCR 0 is read, since this is the only PCR that will actually be used later. This is mainly due to laziness: a single call of `Esys_PCR_Read` can only fetch up to 8 PCRs in total, so we would need to loop over multiple calls to read all available PCRs. This is totally doable, see e.g. the [`tpm2_pcrread` source code](https://github.com/tpm2-software/tpm2-tools/blob/master/tools/tpm2_pcrread.c), but seems a bit unnecessary in this case.
- [ ] **I didn't test this at all with the fwupd binary because I don't own compatible hardware.** The PCR reading works as expected, I tested it in a standalone program, the code that prints the hash and adds it to the list might be buggy though. The CI build [passes successfully](https://travis-ci.com/diabonas/fwupd/builds/124576964).
- [ ] [`.lgtm.yml`](https://github.com/fwupd/fwupd/blob/master/.lgtm.yml) needs to be updated to install tpm2-tss from source, see e.g. the [tpm2-tools `.lgtm.yml`](https://github.com/tpm2-software/tpm2-tools/blob/master/.lgtm.yml) for inspiration.

superm1 commented:
Thanks for doing this!  I think it's a much better solution.

> Currently only PCR 0 is read, since this is the only PCR that will actually be used later. This is mainly due to laziness: a single call of Esys_PCR_Read can only fetch up to 8 PCRs in total, so we would need to loop over multiple calls to read all available PCRs. This is totally doable, see e.g. the tpm2_pcrread source code, but seems a bit unnecessary in this case.

I don't see a problem with this.

> I didn't test this at all with the fwupd binary because I don't own compatible hardware. The PCR reading works as expected, I tested it in a standalone program, the code that prints the hash and adds it to the list might be buggy though. The CI build passes successfully.

@hughsie considering it's a new hard dep, should we make this configurable?

> .lgtm.yml needs to be updated to install tpm2-tss from source, see e.g. the tpm2-tools .lgtm.yml for inspiration.

Snap will also need to be updated.  I think we can clean up lgtm and circle CI based stuff (like snap) after this is approved and merged though.

superm1 commented on contrib/debian/control.in:
```diff
@@ -34,8 +34,7 @@ Depends: ${misc:Depends},
 Recommends: python3,
 	    bolt,
-	    tpm2-tools,
-	    tpm2-abrmd,
```
If i'm not mistaken the TSS library still will use tpm2-abrmd if installed right?  So probably want to keep tpm2-abrmd as a Recommends at least.

diabonas replied:
It does use tpm2-abrmd if installed, yes. When using the Enhanced System API (ESAPI) like we do in this PR, I don't think tpm2-abrmd is really necessary because it will default to using the in-kernel resource manager (`/dev/tpmrm0`) available since kernel 4.12 anyway. Since tpm2-abrmd is generally useful for more complex application scenarios, I don't mind keeping it though, it is only in "recommended" anyway.

diabonas replied:
Yeah, non-exclusive TPM access shouldn't be a problem with recent kernels. Come to think of it, there is another possible benefit of tpm2-abrmd though: if it is installed, ordinary users should be able to access the TPM out of the box without needing super user rights. The in-kernel resource manager OTOH is normally owned by the `tss` group, so users need to be manually added to this group to gain access. This might not be a concern for fwupd because it is probably run as root anyway.

diabonas replied:
OK, I dropped tpm2-abrmd for now, let me know if it needs to be added back.

superm1 commented on contrib/fwupd.spec.in:
```diff
@@ -106,7 +106,7 @@ Requires: bubblewrap
 
 Recommends: python3
-Recommends: tpm2-tools tpm2-abrmd
+Recommends: tpm2-tss
```
Likewise same comment on the debian packaging.

diabonas replied:
Done, see above.

superm1 commented on plugins/uefi/fu-uefi-pcrs.c:
```diff
@@ -103,10 +96,52 @@ fu_uefi_pcrs_setup_tpm12 (FuUefiPcrs *self, const gchar *fn_pcrs, GError **error
+		return FALSE;
+	}
+
+	/* currently we only use PCR 0 on banks SHA-1 and SHA-256, so fetch only these */
```
Who is we in this context?  Do you mean generally only SHA1 and SHA256 are used for PCR0?
I would wonder if we can future proof this a little further for SHA384 and SHA512.  Does tpm2-tss support those yet?

superm1 commented on plugins/uefi/fu-uefi-pcrs.c:
```diff
@@ -103,10 +96,52 @@ fu_uefi_pcrs_setup_tpm12 (FuUefiPcrs *self, const gchar *fn_pcrs, GError **error
+	                    &pcr_selection_in, NULL, NULL, &pcr_values);
+	Esys_Finalize (&ctx);
+	if (rc != TSS2_RC_SUCCESS) {
+		return FALSE;
```
In this case `error` needs to be filled out.

diabonas replied:
Done.

superm1 commented on plugins/uefi/fu-uefi-pcrs.c:
```diff
@@ -103,10 +96,52 @@ fu_uefi_pcrs_setup_tpm12 (FuUefiPcrs *self, const gchar *fn_pcrs, GError **error
+
+	rc = Esys_Initialize (&ctx, NULL, NULL);
+	if (rc != TSS2_RC_SUCCESS) {
+		return FALSE;
```
likewise `error` needs to be filled out here.

superm1 commented on plugins/uefi/fu-uefi-pcrs.c:
```diff
@@ -103,10 +96,52 @@ fu_uefi_pcrs_setup_tpm12 (FuUefiPcrs *self, const gchar *fn_pcrs, GError **error
+	TSS2_RC rc;
+	ESYS_CONTEXT *ctx = NULL;
+	TPML_PCR_SELECTION pcr_selection_in = { 0, };
+	TPML_DIGEST *pcr_values = NULL;
```
Would it be possible to do a custom `g_autoptr` for `TPML_DIGEST` and `ESYS_CONTEXT` to make cleanup and error handling better in this function?  More information about this stuff here: https://developer.gnome.org/glib/stable/glib-Miscellaneous-Macros.html#G-DEFINE-AUTOPTR-CLEANUP-FUNC:CAPS

diabonas replied:
Done, added `G_DEFINE_AUTOPTR_CLEANUP_FUNC` for `ESYS_CONTEXT` and used `g_autofree` for `TPML_DIGEST`.

superm1 commented on plugins/uefi/fu-uefi-pcrs.c:
```diff
@@ -103,10 +96,52 @@ fu_uefi_pcrs_setup_tpm12 (FuUefiPcrs *self, const gchar *fn_pcrs, GError **error
+	ESYS_CONTEXT *ctx = NULL;
+	TPML_PCR_SELECTION pcr_selection_in = { 0, };
+	TPML_DIGEST *pcr_values = NULL;
+	g_autoptr(GString) str = NULL;
```
It looks to me this can be moved into the `for` loop.

superm1 commented on plugins/uefi/fu-uefi-pcrs.c:
```diff
@@ -103,10 +96,52 @@ fu_uefi_pcrs_setup_tpm12 (FuUefiPcrs *self, const gchar *fn_pcrs, GError **error
+		g_debug ("added PCR-%02u=%s", item->idx, item->checksum);
+	}
+
+	g_free (pcr_values);
```
This can be removed if you can setup an autoptr cleanup function.

superm1 commented on plugins/uefi/fu-self-test.c:
```diff
@@ -36,57 +36,6 @@ fu_uefi_pcrs_1_2_func (void)
-	pcrXs = fu_uefi_pcrs_get_checksums (pcrs, 999);
-	g_assert_nonnull (pcrXs);
-	g_assert_cmpint (pcrXs->len, ==, 0);
-}
```
Do you guys have any instrumentation methods to use with the tss stack?  It would be good to keep a test case here, even if it's using a simulated TPM or instrumented data.

diabonas replied:
We usually start an instance of the [IBM TPM simulator](https://sourceforge.net/projects/ibmswtpm2/) to run tests against. tpm2-tss has support for this out of the box, so if you start `tpm_server` in the background and have no physical TPM available, the simulator will be used. Hence we could use something like this as a test if a physical or simulated TPM is available:
```
static void
fu_uefi_pcrs_2_0_func (void)
{
	gboolean ret;
	g_autoptr(FuUefiPcrs) pcrs = fu_uefi_pcrs_new ();
	g_autoptr(GError) error = NULL;
	g_autoptr(GPtrArray) pcr0s = NULL;
	g_autoptr(GPtrArray) pcrXs = NULL;

	/* unset FWUPD_SYSFSTPMDIR so that fu_uefi_pcrs_setup_tpm20 instead of
	   fu_uefi_pcrs_setup_tpm12 is tested */
	g_unsetenv ("FWUPD_SYSFSTPMDIR");

	ret = fu_uefi_pcrs_setup (pcrs, &error);
	g_assert_no_error (error);
	g_assert_true (ret);
	pcr0s = fu_uefi_pcrs_get_checksums (pcrs, 0);
	g_assert_nonnull (pcr0s);
	g_assert_cmpint (pcr0s->len, >=, 2);
	pcrXs = fu_uefi_pcrs_get_checksums (pcrs, 999);
	g_assert_nonnull (pcrXs);
	g_assert_cmpint (pcrXs->len, ==, 0);
}
```
However I don't know how to guarantee that the simulator is started: I tried using [`g_spawn_async`](https://developer.gnome.org/glib/stable/glib-Spawning-Processes.html#g-spawn-async), but that is suboptimal since a failed assertion means that the program is terminated and doesn't kill its child process (and there seems to be no native kill function in GLib anyway, you need to use `kill()` directly).

diabonas replied:
Done, I went for the Arch Linux CI, since that saves me from building the TPM simulator from source and added the above test. It is run if physical TPM is found or`TPM_SERVER_RUNNING` is set, otherwise it is skipped.

diabonas commented on contrib/debian/control.in:
```diff
@@ -34,7 +34,7 @@ Depends: ${misc:Depends},
 Recommends: python3,
 	    bolt,
-	    tpm2-tools,
+	    tpm2-tss,
```
OK cool, I simply removed the tpm2-tools dependency for Debian and Fedora, hoping that it will work automagically (or be fixed by someone more knowledgeable in packaging for these distributions) ;)

diabonas commented on contrib/fwupd.spec.in:
```diff
@@ -106,7 +106,7 @@ Requires: bubblewrap
 
 Recommends: python3
-Recommends: tpm2-tools tpm2-abrmd
+Recommends: tpm2-tss tpm2-abrmd
```
Removed tpm2-tools and didn't add tpm2-tss for now.

diabonas commented on plugins/uefi/fu-uefi-pcrs.c:
```diff
@@ -103,10 +95,74 @@ fu_uefi_pcrs_setup_tpm12 (FuUefiPcrs *self, const gchar *fn_pcrs, GError **error
+	rc = Esys_GetCapability (ctx, ESYS_TR_NONE, ESYS_TR_NONE, ESYS_TR_NONE,
+	                         TPM2_CAP_PCRS, 0, 1, NULL, &capability_data);
+	if (rc != TSS2_RC_SUCCESS) {
+		Esys_Finalize (&ctx);
```
Good idea, done.

diabonas commented on plugins/uefi/fu-uefi-pcrs.c:
```diff
@@ -103,10 +95,74 @@ fu_uefi_pcrs_setup_tpm12 (FuUefiPcrs *self, const gchar *fn_pcrs, GError **error
+
+	rc = Esys_PCR_Read (ctx, ESYS_TR_NONE, ESYS_TR_NONE, ESYS_TR_NONE,
+	                    &pcr_selection_in, NULL, NULL, &pcr_values);
+	Esys_Finalize (&ctx);
```
Done, see above.

diabonas commented on plugins/uefi/fu-self-test.c:
```diff
@@ -39,54 +41,36 @@ fu_uefi_pcrs_1_2_func (void)
-		  "  0  : 122de8b579cce17b0703ca9f9716d6f99125af9569e7303f51ea7f85d317f01e\n"
-		  /* new output format of tpm2_pcrread */
-		  "  1 : 0x0D89E7CA2C487EED36DB5684826B4ABF0904D3BDCD74E61999573570CF3F9C75\n", TRUE);
+	g_unsetenv ("FWUPD_SYSFSTPMDIR");
```
Unfortunately yes because the environment variables are not scoped: `FWUPD_SYSFSTPMDIR` is still set from `fu_uefi_pcrs_1_2_func`, so `fu_uefi_pcrs_setup` will find a TPM 1.2 device if the variable is not unset first. I just moved the environment variable definition from `fu_uefi_ucs2_func` to `fu_uefi_pcrs_1_2_func` so that the order in which `fu_uefi_pcrs_1_2_func` and `fu_uefi_pcrs_2_0_func` are run does not matter.

hughsie replied:
or unset it in the TPM 1.2 function?

diabonas replied:
Ah yes, that is cleaner indeed. I was worried first that `g_unsetenv` might be skipped if an assertion in `fu_uefi_pcrs_1_2_func` fails, but in that case complete test program is aborted, so this approach should be safe.

diabonas commented:
> Here is what should do the trick for the snap stuff if you can include it.

Great, I added it to the PR, thanks!



> So I'm noticing this on Ubuntu eoan with this branch right now with output debugging enabled.
> 
> ```
> WARNING:esys:src/tss2-esys/esys_tcti_default.c:137:tcti_from_file() Could not load TCTI file: libtss2-tcti-default.so 
> WARNING:esys:src/tss2-esys/esys_tcti_default.c:137:tcti_from_file() Could not load TCTI file: libtss2-tcti-tabrmd.so 
> ```

These warnings are annoying but harmless: the auto-detection which library should be used to access the TPM goes through different libraries and uses the first one that it finds, but throws these warnings for the non-existent ones in the process. The only way to suppress them for tpm2-tss <2.3 is to either set the default library by symlinking `/usr/lib/libtss2-tcti-device.so` to `/usr/lib/libtss2-tcti-default.so` or to export the environment variable `TSS2_LOG=esys+error` so that only real errors are displayed.

The most recent version 2.3.0 doesn't show these messages any more, see https://github.com/tpm2-software/tpm2-tss/pull/1480, but it might take a while until it gets packaged for Ubuntu.

diabonas commented:
> > export the environment variable `TSS2_LOG=esys+error` so that only real errors are displayed.
> 
> This sounds fine to me. We can set it only if running fwupd in non-verbose mode. In verbose mode show this kind of stuff.

I have added this environment variable to `plugins/uefi/fu-uefi-tool.c` when in non-verbose mode, does this suppress the warnings on Ubuntu?

Furthermore, I added tpm2-tss to the LGTM configuration.

diabonas commented:
@superm1 I applied the suggested change

hughsie commented on plugins/uefi/fu-uefi-pcrs.c:
```diff
@@ -103,10 +101,77 @@ fu_uefi_pcrs_setup_tpm12 (FuUefiPcrs *self, const gchar *fn_pcrs, GError **error
+			capability_data->data.assignedPCR.pcrSelections[i].hash;
+		pcr_selection_in.pcrSelections[i].sizeofSelect =
+			capability_data->data.assignedPCR.pcrSelections[i].sizeofSelect;
+		pcr_selection_in.pcrSelections[i].pcrSelect[0] = 0b00000001;
```
Holy cow, how long have `0b` prefixes been a thing?

hughsie commented:
I'm really impressed with the quality of this PR, great job guys!
